### ISSUE TITLE
Privatization test fails on mpi-smp

### ISSUE THREAD

**Issue by nitbhat** [UIUC-PPL/charm#3274]
Autobuild log: https://charm.cs.illinois.edu/autobuild/old.2021_02_04__01_08/mpi-linux-x86_64-smp.txt

```
../../../bin/ampif90 -optimize -production -o pipglobals-test-f90.o -pipglobals -c test-f90.f90
../../../bin/ampif90 -optimize -production -o f90-pipglobals -pipglobals pipglobals-test.o pipglobals-test-f90.o -balancer TreeLB
Note: PiP-glibc not found. Maximum virtualization count will be limited.
      Please set and export $PIP_GLIBC_INSTALL_DIR appropriately.
../../../bin/testrun  ./f90-pipglobals +p1 +vp2 +balancer RandCentLB ++local +CmiSleepOnIdle +setcpuaffinity

Running on 1 processors:  ./f90-pipglobals +vp2 +balancer RandCentLB +CmiSleepOnIdle +setcpuaffinity 
charmrun>  /usr/bin/setarch x86_64 -R  mpirun -np 1  ./f90-pipglobals +vp2 +balancer RandCentLB +CmiSleepOnIdle +setcpuaffinity 
Charm++> Running on MPI version: 3.0
Charm++> level of thread support used: -1 (desired: 0)
Charm++> Running in SMP mode: 1 processes, 1 worker threads (PEs) + 1 comm threads per process, 1 PEs total
Charm++> The comm. thread both sends and receives messages
Converse/Charm++ Commit ID: 9ea7be9
Charm++ built with internal error checking enabled.
Do not use for performance benchmarking (build without --enable-error-checking to do so).
CharmLB> Load balancer assumes all CPUs are same.
Charm++> cpu affinity enabled. 
Charm++> Running on 1 hosts (2 sockets x 4 cores x 1 PUs = 8-way SMP)
Charm++> cpu topology info is gathered in 0.000 seconds.
[0] TreeLB in LEGACY MODE support
[0] TreeLB: Using PE_Root tree with: Random 
AMPI> Using pipglobals privatization method.
Beginning round of testing.
Test #01: module variable
#01 - [0](0) - 0x7ffff1a780a4 - single write - passed
#01 - [1](0) - 0x7ffff186a0a4 - single write - passed
#01 - [0](0) - 0x7ffff1a780a4 - many write - passed
#01 - [1](0) - 0x7ffff186a0a4 - many write - passed
Test #02: implicit save variable
#02 - [0](0) - 0x7ffff1a780ac - single write - passed
#02 - [1](0) - 0x7ffff186a0ac - single write - passed
#02 - [0](0) - 0x7ffff1a780ac - many write - passed
#02 - [1](0) - 0x7ffff186a0ac - many write - passed
Test #03: explicit save variable
#03 - [0](0) - 0x7ffff1a780a8 - single write - passed
#03 - [1](0) - 0x7ffff186a0a8 - single write - passed
#03 - [0](0) - 0x7ffff1a780a8 - many write - passed
#03 - [1](0) - 0x7ffff186a0a8 - many write - passed
Test #04: subroutine save variable
#04 - [0](0) - 0x7ffff1a780b0 - single write - passed
#04 - [1](0) - 0x7ffff186a0b0 - single write - passed
#04 - [0](0) - 0x7ffff1a780b0 - many write - passed
#04 - [1](0) - 0x7ffff186a0b0 - many write - passed
Test #05: common block variable
#05 - [0](0) - 0x7ffff1a79b50 - single write - passed
#05 - [1](0) - 0x7ffff186bb50 - single write - passed
#05 - [0](0) - 0x7ffff1a79b50 - many write - passed
#05 - [1](0) - 0x7ffff186bb50 - many write - passed
Round of testing complete.
Beginning round of testing.
Test #06: module variable
#06 - [0](0) - 0x7ffff1a780a4 - single write - passed
#06 - [1](0) - 0x7ffff186a0a4 - single write - passed
#06 - [0](0) - 0x7ffff1a780a4 - many write - passed
#06 - [1](0) - 0x7ffff186a0a4 - many write - passed
Test #07: implicit save variable
#07 - [0](0) - 0x7ffff1a780ac - single write - passed
#07 - [1](0) - 0x7ffff186a0ac - single write - passed
#07 - [0](0) - 0x7ffff1a780ac - many write - passed
#07 - [1](0) - 0x7ffff186a0ac - many write - passed
Test #08: explicit save variable
#08 - [0](0) - 0x7ffff1a780a8 - single write - passed
#08 - [1](0) - 0x7ffff186a0a8 - single write - passed
#08 - [0](0) - 0x7ffff1a780a8 - many write - passed
#08 - [1](0) - 0x7ffff186a0a8 - many write - passed
Test #09: subroutine save variable
#09 - [0](0) - 0x7ffff1a780b0 - single write - passed
#09 - [1](0) - 0x7ffff186a0b0 - single write - passed
#09 - [0](0) - 0x7ffff1a780b0 - many write - passed
#09 - [1](0) - 0x7ffff186a0b0 - many write - passed
Test #10: common block variable
#10 - [0](0) - 0x7ffff1a79b50 - single write - passed
#10 - [1](0) - 0x7ffff186bb50 - single write - passed
#10 - [0](0) - 0x7ffff1a79b50 - many write - passed
#10 - [1](0) - 0x7ffff186bb50 - many write - passed
Round of testing complete.
All tests passed.
[Partition 0][Node 0] End of program
------------- Processor 1 Exiting: Caught Signal ------------
Reason: Segmentation fault
[1] Stack Traceback:
  [1:0] f90-pipglobals 0x7439d8 
  [1:1] libpthread.so.0 0x7ffff7bcb390 
  [1:2] libpthread.so.0 0x7ffff144822e 
*** The MPI_Abort() function was called after MPI_FINALIZE was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[might:27370] Local abort after MPI_FINALIZE completed successfully; not able to aggregate error messages, and not able to guarantee that all other processes were killed!
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code.. Per user-direction, the job has been aborted.
-------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[9201,1],0]
  Exit code:    1
--------------------------------------------------------------------------

real	0m0.166s
user	0m0.064s
sys	0m0.066s
Makefile:157: recipe for target 'test-f90-pipglobals' failed
make[2]: *** [test-f90-pipglobals] Error 1
make[2]: Target 'test' not remade because of errors.
rm cxx-pipglobals-shared-library-so.o cxx-tlsglobals-static-library.o cxx-fsglobals-static-library.o cxx-pipglobals-static-library.o cxx-tlsglobals-shared-library-so.o
make[2]: Leaving directory '/scratch/autobuild/mpi64-smp/charm/mpi-linux-x86_64-smp/tests/ampi/privatization'
Makefile:52: recipe for target 'test-privatization' failed
make -C exit test OPTS='-optimize -production' TESTOPTS='++local +CmiSleepOnIdle +setcpuaffinity'
make[1]: *** [test-privatization] Error 2
```

**Comment by evan-charmworks** [comment]
I think I recognize this issue (crash at exit with pipglobals) from some experiments I've done.

**Comment by evan-charmworks** [comment]
#3287 on its own does not resolve the issue.

Important details include:
- the backtrace of the segfault shows it happens inside the implementation of `dlclose`
- the test only crashes on MPI SMP, where the comm thread is thread 0 instead of thread N, and so dlmopen will not be called on thread 0
- the test succeeds for C++ but fails for F90 (implying libgfortran is involved)
- the test crashes even with `+p1`
- the worker thread is not fully exited when the crash happens, it is in a loop calling sched_yield from Converse exit / realCkExit

**Comment by stwhite91** [comment]
It sounds like we need to ensure that dlmopen is only called on thread 0 then. Even with +p1 you'll get a commthread on thread 0 in MPI-SMP. And we don't want to change that behavior of our MPI-SMP build since it is required to use MPI_THREAD_FUNNELED and not MPI_THREAD_SERIALIZED or MPI_THREAD_MULTIPLE, both of which can have higher overheads than FUNNELED.

**Comment by evan-charmworks** [comment]
Fixed by #3287
